File: ffi/python/blsct/tx_key.py
import blsct
from typing import Any
from ..child_key import ChildKey
from ..scalar_based_key import ScalarBasedKey
from tx_key_derived.spending_key import SpendingKey
from tx_key_derived.view_key import ViewKey

class TxKey(ScalarBasedKey):
  def __init__(self, child_key: ChildKey):
    # tx_key is of type *BlsctScalar
    tx_key = blsct.from_child_key_to_tx_key(child_key.value())
    super().__init__(tx_key)

  def to_spending_key(self) -> SpendingKey:
    return SpendingKey(self)

  def to_view_key(self) -> ViewKey:
    return ViewKey(self)

File: ffi/ts/src/keys/childKeyDesc/txKey.ts
import {
  fromTxKeyToSpendingKey,
  fromTxKeyToViewKey
} from '../../blsct'

import { Scalar } from '../../scalar'
import { SpendingKey } from './txKeyDesc/spendingKey'
import { ViewKey } from './txKeyDesc/viewKey'

/** Represents a tx key. A tx key is a Scalar and introduces no new functionality; it serves purely as a semantic alias. Both SpendingKey and ViewKey are exclusively derived from a TxKey.
 *
 * Examples:
 * ```ts
 * const { Scalar, ChildKey } = require('navio-blsct')
 * const s = Scalar.random()
 * const ck = new ChildKey(s)
 * const txk = ck.toTxKey()
 * txk.toSpendingKey()
 * txk.toViewKey()
 * ```
 */
export class TxKey extends Scalar {
  constructor(obj?: any) {
    super(obj)
  }

  toSpendingKey(): SpendingKey {
    const obj = fromTxKeyToSpendingKey(this.value())
    return new SpendingKey(obj)
  }

  toViewKey(): ViewKey {
    const obj = fromTxKeyToViewKey(this.value())
    return new ViewKey(obj)
  }
}

